kotest-property/kotest-property-core/src/commonMain/kotlin/io/kotest/property/core/proptest.kt

```diff
@@ -0,0 +1,59 @@
+package io.kotest.property.core
+
+import io.kotest.mpp.sysprop
+import io.kotest.property.AssumptionFailedException
+import io.kotest.property.Gen
+import io.kotest.property.PropertyTesting
+import io.kotest.property.RandomSource
+import io.kotest.property.ShrinkingMode
+import kotlin.random.Random
+import kotlin.time.Duration
+
+class PropTestContext {
+
+   var iterations: Int? = null // converted to iteration based constraint
+   var duration: Duration? = null // converted to duration based constraint
+   var constraints: Constraints? = null // specify manual constraits
+   var minSuccess = -1
+   var maxFailure = -1
+   var edgecaseFactor = PropertyTesting.defaultEdgecasesGenerationProbability
+   var shouldPrintShrinkSteps = false // sysprop("kotest.proptest.output.shrink-steps", true)
+   var shrinkingMode: ShrinkingMode = PropertyTesting.defaultShrinkingMode
+   var maxDiscardPercentage: Int = 20
+   var shouldPrintGeneratedValues: Boolean = PropertyTesting.shouldPrintGeneratedValues
+
+   // must call seed before rs otherwise has no effect
+   var seed: Long? = null
+   var failOnSeed: Boolean = sysprop("kotest.proptest.seed.fail-if-set", false)
+
+   internal val rs = lazy { RandomSource.seeded(seed ?: Random.nextLong()) }
+
+   internal val container = GenDelegateContainer()
+
+   internal var property: suspend () -> Unit = {}
+   internal var beforeProperty: suspend () -> Unit = {}
+   internal var afterProperty: suspend () -> Unit = {}
+
+   fun assume(predicate: () -> Boolean) {
+
+   }
+
+   fun property(fn: suspend () -> Unit) {
+      property = fn
+   }
+
+   fun beforeProperty(fn: suspend () -> Unit) {
+      beforeProperty = fn
+   }
+
+   fun afterProperty(fn: suspend () -> Unit) {
+      afterProperty = fn
+   }
+
+   fun <T> gen(initializer: () -> Gen<T>): GenDelegate<T> {
```

Comment:
Suggestion to allow direct delegation of generators:

```kotlin
override operator <A> fun Gen<A>.getValue(thisRef: Any?, property: KProperty<*>): A {
  // ...
}
```

Didn't study all the bits and pieces of GenDelegate though, so maybe there's a good reason to keep it around.

Reply:
I think if we want to allow the delegates to be declared inside the test block (which is now the only place it can be declared, see proposal 2), then we need this extra "hop" to avoid the delegate being re-created each time, and the stream of values reset.

Reply:
In proposal 2, withPermutuation{} starts the test immediately.

Reply:
> then we need this extra "hop" to avoid the delegate being re-created each time

In its current form that's true. I'm wondering though if it has to be that way - after all it's just a function call, so there might be a way to keep the actual generator object around. Might be tricky though.

But then, how does `gen { }` even help? If cross-dependencies are possible and generator B depends on A, you certainly need to recreate B every time. So unless there is some compiler magic detecting that, it's currently either always created for every iteration, or cross-dependencies are simply not working. Am I missing anything?

However, maybe the best approach is the most straight-forward one:

- Allow cross-dependencies, get rid of `gen { }`
- Any generator that does not have dependencies on generated data, you can simply define outside the block (and then assign via by inside the block)

Downside: two variables for one thing